src/handles/effect.ts

```diff
@@ -0,0 +1,133 @@
+/** @noSelfInFile **/
+
+import { Handle } from "./handle";
+import { MapPlayer } from "./player";
+import { Point } from "./point";
+
+export class Effect extends Handle<effect> {
+
+  constructor(m: string, s: number | widget, t: number | string) {
+    if (typeof s == "number" && typeof t == "number") {
+      super(AddSpecialEffect, [m, s, t]);
+    // TODO: Find a better solution, if possible.
+    } else {
+      super(AddSpecialEffectTarget, [m, s, t]);
+    }
+  }
```

Comment:
I think the constructor would make much more sense to be written this way:

```typescript
  constructor(m: string, x: number, y: number);
  constructor(m: string, targetWidget: Widget, attachPointName: string);
  constructor(m: string, a: number | Widget, b: number | string) {
    if (typeof a == "number") {
      super(AddSpecialEffect, [m, a, b]);
    } else {
      super(AddSpecialEffectTarget, [m, a.handle, b]);
    }
  }
```

In your current setup an effect can be constructed with a widget but the user can accidentally use a number instead of a string which is invalid for AddSpecialEffectTarget, and vice versa.



We also want to use the Widget wrapper type and not the native widget handle type. This way users won't have to pass `unit.handle`, instead they can just pass `unit`.

Reply:
As I said on discord, my TypeScript wont compile that because unit is not an object of type Widget. I don't see how it could be, being Unit not extending Widget.

Reply:
Take this simplified code for example.

```typescript
declare abstract class handle { __handle: never; }
declare interface agent extends handle { __agent: never; }
declare interface widget extends agent { __widget: never; }
declare interface unit extends widget { __unit: never; }
declare interface group extends agent { __group: never; }

export class Handle<T extends handle> {

  private handleVar: T;
  protected static initHandle: handle | undefined;

  protected constructor(initFunc?: (...args: any) => any, initArgs?: any[]) {
    this.handleVar = Handle.initHandle as T;
  }

  public get handle(): T {
    return this.handleVar;
  }

}

export class Widget extends Handle<widget> {
}

export class Unit extends Handle<unit> {
    constructor() { 
        super();
    }
}

export class Group extends Handle<group> {
    constructor() {
        super();
    }
}

function DoSomethingWithWidget(u: Widget) {
}

// valid
const u = new Unit();
DoSomethingWithWidget(u);

// invalid
const g = new Group();
DoSomethingWithWidget(g);

/**
  Argument of type 'Group' is not assignable to parameter of type 'Widget'.
  Types of property 'handleVar' are incompatible.
    Property '__widget' is missing in type 'group' but required in type 'widget'.
 */
```

In order for Unit to be used as a Widget the Unit class must have all the properties of Widget.  This also includes the `handleVar` property of the base Handle class. Since `unit` inherits `widget`, `handleVar` will have the `__widget` property. As you can see in the example, Group's signature does not match because the `group` type is missing the `__widget` property (from the `handleVar` instance variable). This is valid TypeScript code, and you can test it by using the TypeScript playground.

https://www.typescriptlang.org/play/

It is somewhat confusing and if you have a better solution I'm open to hear it.

Reply:
I realized I can make Unit extend Widget by overriding the handle property. 
```typescript
export class Unit extends Widget {
  readonly handle: unit;
}
```
However it will require some changes to the base handle class and in turn slightly changing how handle constructors will work. If you want to just use `widget` instead of `Widget` temporarily in this PR I can update it when I update the base handle class.

Reply:
I'm not at home right now but I'll do that once I am.

Reply:
Should be good for now until `Widget` is "fixed".